examples/checkpointing/checkpointing_example.py

```diff
@@ -0,0 +1,142 @@
+import numpy as np
+from numpy import linalg
+from math import floor
+from devito import Function, TimeFunction
+from examples.seismic import Receiver, demo_model, RickerSource
+from examples.seismic.acoustic import ForwardOperator, GradientOperator, smooth10
+from examples.checkpointing.checkpoint import DevitoCheckpoint, CheckpointOperator
+from pyrevolve import Revolver
+
+
+def setup(shape, tn, spacing, time_order, space_order, nbpml):
+    nrec = shape[0]
+    model = demo_model('layers-isotropic', shape=shape, spacing=spacing, nbpml=nbpml)
+    dt = model.critical_dt * (1.73 if time_order == 4 else 1.0)
+    t0 = 0.0
+    nt = int(1 + (tn-t0) / dt)  # Number of timesteps
+    time = np.linspace(t0, tn, nt)  # Discretized time axis
+
+    # Define source geometry (center of domain, just below surface)
+    src = RickerSource(name='src', grid=model.grid, f0=0.01, time=time)
+    src.coordinates.data[0, :] = np.array(model.domain_size) * .5
+    src.coordinates.data[0, -1] = model.origin[-1] + 2 * spacing[-1]
+
+    # Define receiver geometry (spread across x, just below surface)
+    # We need two receivers - one for the true (verification) run
+    rec_t = Receiver(name='rec_t', grid=model.grid, ntime=nt, npoint=nrec)
+    rec_t.coordinates.data[:, 0] = np.linspace(0., model.domain_size[0], num=nrec)
+    rec_t.coordinates.data[:, 1:] = src.coordinates.data[0, 1:]
+
+    # and the other for the smoothed run
+    rec_s = Receiver(name='rec_s', grid=model.grid, ntime=nt, npoint=nrec,
+                     coordinates=rec_t.coordinates.data)
+
+    # Receiver for Gradient run
+    rec_g = Receiver(name="rec_g", grid=model.grid, ntime=nt,
+                     coordinates=rec_s.coordinates.data)
+
+    # Create the forward wavefield to use (only 3 timesteps)
+    u = TimeFunction(name="u", grid=model.grid, time_order=time_order,
+                     space_order=space_order, save=False, dtype=model.dtype)
```

Comment:
`dtype` parameter is no longer needed, as it's inferred from `model`.

Reply:
Fixed